obs2ioda alignment

BenjaminRuston opened the pull request:
## Description

this brings in updated obs2ioda which can now be aligned and maintained as a unified repository incorporating bug fixes from all developers


## Issue(s) addressed

Resolves #1670 



## Impact

single unified obs2ioda repository

jamiebresch committed:
add obs2ioda (legacy v1 format) converter

Moved from ioda-converters/src/ncar-bufr2nc-fortran and extended to convert Hiwamari Standard Data.

jamiebresch committed:
Update README.md

jamiebresch committed:
rename directory

jamiebresch committed:
add preliminary code for converting satwnd bufr file

jamiebresch committed:
add flags for writing out selected MetaData

jamiebresch committed:
Update README.md

jamiebresch committed:
Update README.md

jamiebresch committed:
add -split option to write out hourly files

for now, hard-coded for -3h/+3h hourly assuming file date at 0h

jamiebresch committed:
update sample Makefile

jamiebresch committed:
add script to convert CISL RDA archived NCEP BUFR files on Cheyenne

jamiebresch committed:
Update README.md

jamiebresch committed:
add AVHRR and VIIRS satwnd processing

also remove unused lines in Makefile

jamiebresch committed:
add LEOGEO satwnd processing

jamiebresch committed:
assign quality marker=2 (neutral or not checked) to data from satwnd.bufr

jamiebresch committed:
add IASI radiance processing

jamiebresch committed:
add CrIS radiance processing

jamiebresch committed:
Update README.md

jamiebresch committed:
Update README.md

jamiebresch committed:
add GOES ABI radiance processing

jamiebresch committed:
fix for nchannel=399 cris.bufr file

jamiebresch committed:
update sample script for IASI and CrIS

jamiebresch committed:
Update README.md

jamiebresch committed:
Update README.md

jamiebresch committed:
sync gnssro code with the one in JCSDA ioda-converters

in addition, change declaration of qfro to qfro(1) to eliminate a compilation warning

jamiebresch committed:
minor update to fgat

jamiebresch committed:
make a copy from obs2ioda-v1 src

jamiebresch committed:
write out ioda-v2 files (exclude gnssro)

jamiebresch committed:
remove char datetime in v2 output

jamiebresch committed:
for obs2ioda-v2, fix units attribute and add nchans(nchans) output

jamiebresch committed:
fix solar_azimuth_angle and solar_zenith_angle output

ibanos90 committed:
Assign IASI observation errors based on values from GSI global_satinfo.txt (#1)

This PR assigns brightness temperature observation errors to IASI observations using the errors provided in the file https://github.com/comgsi/fix/blob/master/global_satinfo.txt which is used in GSI.

**Changes**:
- A separate subroutine is added to include the assignment of observation errors of hyper-spectral sensors.

**Tests**:
- Tested both obs2ioda-v1 and obs2ioda-v2 and converted files contain the added information;
- Tests with Hofx complete successfully with some filters.

jamiebresch committed:
fix slowness in setting Tb obs errors

jamiebresch committed:
add sensor_band_central_radiation_wavenumber output for IASI and CrIS

VarMetaData group is deprecated in ioda-v2, but it is used in some UFO code.
For now still write out sensor_band_central_radiation_wavenumber in VarMetaData group.

jamiebresch committed:
fix typo in declaring xinfo_int and xinfo_int64

jamiebresch committed:
sync gnssro code with the one in JCSDA ioda-converters

jamiebresch committed:
apply valid_range attribute fixes for gnssro as in JCSDA ioda-converters

See https://github.com/JCSDA-internal/ioda-converters/pull/780

ibanos90 committed:
Add superobbing for GOES ABI data

ibanos90 committed:
Add comment on the superobbing code followed

ibanos90 committed:
Update README.md

ibanos90 committed:
Modify ahi reading and bugs to write the ioda files

ibanos90 committed:
Fix reading of data for radiance calculation

ibanos90 committed:
Use '&' when reading the variable for better code readability. Add some comments for documentation

ibanos90 committed:
Fixing bug from previous commit. The reading of idata was deleted unintentionally

ibanos90 committed:
Add superobbing for AHI data

ibanos90 committed:
Merge pull request #3 from NCAR/feature/ahiFix

**Description**
This PR fixes some bugs found to convert raw AHI observations from the AWS cloud in Himawari Standard Data (HSD) format to the IODA-v2 needed in `mpas-jedi`. It was added the capability to read files with segment number 01, which corresponds to the no segment division of HSD data. The reading of the variables was modified to read each one explicitly since otherwise it was not working properly (not sure why). Specifically the way the variable `idata`, which is the count (total number of pixels that results from multiplying the number of columns by the number of lines) needed for the radiance calculation is also modified. This modification results in the correct calculation of the radiance and then the brightness temperature. In addition, the variable for the `dateTime` in `epochtime` type was allocated since that was missing. The observation error was also added following the values used in GSI (https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/gfs.v16.2.2/fix/fix_gsi/global_satinfo.txt) for clear-sky and a bug in the function that sets the errors was fixed as well.

**Closes** 
Issue https://github.com/NCAR/obs2ioda/issues/2

ibanos90 committed:
Delete unnecessary print

ibanos90 committed:
Merge branch 'main' into feature/ahiSuperob

ibanos90 committed:
Remove debugging build flags

ibanos90 committed:
Merge pull request #4 from NCAR/feature/ahiSuperob

This PR adds the superobbing capability for AHI observations. Giving the half width of the superobbing grid as an entry parameter, the code computes the average brightness temperature in the box defined using that parameter. This code is adapted from https://github.com/wrf-model/WRF/blob/develop/var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc and follows the same implementation as for ABI observations from GOES-16. The debugging build flags for the goes_abi code are removed. The following figures show the brightness temperature results for AHI observations after using this converter for 2018-04-15T00 channel 8 (a) without superobbing and  (b) with superobbing on a grid of 59x59 pixels:
**a)**
![RAW_2018041500_b8coverage_test_20220721_ch8_08_12](https://user-images.githubusercontent.com/33061897/184459639-fa6a39bb-9d83-4458-8724-3b42922898b5.png)
**b)**
![RAW_2018041500_b8coverage_test_20220721_superobbing59X59](https://user-images.githubusercontent.com/33061897/184459648-9ac5d155-76a5-435c-be7a-aa6e5255743e.png)

**Issue closed**
Closes https://github.com/NCAR/obs2ioda/issues/1

zhumingying committed:
Adding obserror for CrIS (#10)

Revised define_mod.f90 with adding obs errors for CrIS-FSR.

For data after 2021: Observation errors for the CrIS-FSR with 431 channels have been added.
For data before 2021: If you are processing CrIS data with 399 channels, you will need to manually add the observation errors.

The obs error assigned here are from the two obs file in UFO (It's easy to copy and paste):
ufo-data/testinput_tier_1/cris-fsr_n20_obs_2020121500_m.nc4
ufo-data/testinput_tier_1/cris-fsr_npp_obs_2018041500_m.nc4

The cris-fsr_npp 's errors in this PR are the same as the following file
https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/gfs.v16.3.16/fix/fix_gsi/global_satinfo.txt
while cris-fsr_n20 's errors are different from global_satinfo.txt.

mrixlam committed:
Modified documentation on converting HSD files (#8)

This PR modifies the documentation on converting the HSD file using obs2ioda. Default value for pixel to skip (option -s) has been changed from 3 to 1 which means no pixel will be skipped.

Modified Files:
M obs2ioda-v1/README.md
M obs2ioda-v2/README.md

amstokely committed:
CMake Build Support for obs2ioda-v2 (#11)

TYPE: enhancement

KEYWORDS: cmake, build system, compiler compatibility

DESCRIPTION OF CHANGES:

The obs2ioda-v2 library lacked support for CMake-based builds, which complicated the build process across different platforms and made compiler compatibility (GNU and Intel) more difficult to manage.

Solution:

1. Added support for building obs2ioda-v2 using CMake, leveraging the modern build system to streamline cross-platform integration.

2. Updated the obs2ioda-v2 README with instructions on how to build the library using CMake and provided additional guidance for building the NCEP BUFR library.

3. Fixed a boolean equality check in hsd.f90 to ensure obs2ioda-v2 compiles successfully on both GNU and Intel compilers.

These changes improve the ease of configuration and build process, while maintaining compatibility with multiple compilers.

amstokely committed:
Added fcheck=all and fbacktrace flags to GNU Debug build. This performs bounds checking and saves additional debug info when obs2ioda is compiled in Debug mode.

amstokely committed:
Fixed segfaults by initializing iloc in AHI file parsing

Prevented out-of-bounds memory access and inaccurate calculations caused by the uninitialized iloc variable, used when parsing AHI files without superobs.

amstokely committed:
Fixed logic for determing which set_obserr subroutine to call in ncio_mod.f90

amstokely committed:
Removed cmake debug flags

amstokely committed:
Merge pull request #17 from NCAR/fix/ahi_iloc

Fix Segmentation Fault and Memory Access Issues in AHI File Parsing by Initializing iloc

amstokely committed:
Merge pull request #18 from NCAR/fix/set_obserr_logic

amstokely committed:
Add standardized compiler flags for GNU and Intel with enhanced debugging options (#21)

This PR introduces standardized Fortran compiler flags tailored for the GNU and Intel compilers, enhancing both build consistency and debugging capability across all build types. The updates are as follows:

GNU Compiler Flags:

General: -ffree-line-length-none (removes line length limits), and -mcmodel=medium (supports larger memory models for larger datasets).
Debug: Aggressive debugging flags including -fbacktrace (provides error backtrace), -ffpe-trap=invalid,zero,overflow (traps floating-point exceptions), and -fcheck=all (enables all runtime checks).
Intel Compiler Flags:

General: -mcmodel=medium (supports larger memory models for larger datasets.)
Debug: Includes -check uninit (checks uninitialized variables), -ftrapuv (traps uninitialized variable usage), -g (produces debug information), -traceback (provides symbolic traceback on errors), and -fpe0 (halts on floating-point exceptions).
These flags standardize the compilation environment, making it easier for developers to catch issues, particularly in Debug builds where aggressive error-trapping is employed.

amstokely committed:
This PR introduces an interface for seamless string and string array interoperability between Fortran and C. (#30)

An example of how this string interface could be used in the context of writing a NetCDF string array variable is shown as follows.
Given the signature of a C/C++ function that writes a NetCDF string array variable:

  int netcdfPutVarString(
            int netcdfID,
            const char *groupName,
            const char *varName,
            const char **data
    );
, which maps to the following fortran iso c binding interface:

  function c_netcdfPutVarString(&
                netcdfID, groupName, varName, data) &
                bind(C, name = "netcdfPutVarString")
            import :: c_int
            import :: c_ptr
            integer(c_int), value, intent(in) :: netcdfID
            type(c_ptr), value, intent(in) :: groupName
            type(c_ptr), value, intent(in) :: varName
            type(c_ptr), value, intent(in) :: data
            integer(c_int) :: c_netcdfPutVarString
        end function c_netcdfPutVarString
, f_c_string_1D_t and f_c_string_t would be used as follows:

function netcdfPutVarString(&
            netcdfID, groupName, varName, data) result(status)
            integer, value, intent(in) :: netcdfID
            character(len=:), allocatable, intent(inout) :: groupName
            character(len=:), allocatable, intent(inout) :: varName
            character(len=:), allocatable, dimension(:), intent(inout)
            type(f_c_string_t) :: f_c_string_groupName
            type(f_c_string_t) :: f_c_string_varName
            type(f_c_string_t_1D) :: f_c_string_data
            integer :: status
            status = C_netcdfPutVarString(&
                netcdfID, &
                f_c_string_groupName(groupName)%to_c(), &
                f_c_string_varName(varName)%to_c(), &
                f_c_string_data(data)%to_c())
        end function c_netcdfPutVarString

amstokely committed:
This commit introduces a new CMake function, obs2ioda_cxx_library, to simplify and standardize the configuration of obs2ioda C++ library targets. (#32)

Function Overview
The obs2ioda_cxx_library function takes a single argument, target, representing the C++ target to configure. It performs the following key actions:

Sets the INSTALL_RPATH property to ensure that shared libraries can be found relative to the target's installation directory.
Retrieves public link libraries by dynamically constructing the variable name ${target}_PUBLIC_LINK_LIBRARIES.
Links the retrieved libraries to the target using target_link_libraries.
This setup ensures proper linkage and runtime configuration for relocatable installations.
Benefits

Improved Modularity: Centralizes and simplifies the configuration of C++ libraries in obs2ioda.
Consistency: Ensures all C++ libraries follow the same setup process for linking and runtime path configuration.
Ease of Use: Reduces repetitive CMake code by encapsulating common configuration steps in a reusable function.

amstokely committed:
Refactor obs2ioda_fortran_target by splitting it into two functions: obs2ioda_fortran_library and obs2ioda_fortran_executable. (#31)

Previously, the function configured both Fortran libraries and their corresponding executables, limiting flexibility. This update introduces separate functions for libraries and executables, removing their dependency and enabling standalone library creation. These changes improve modularity, clarify function use, and enhance maintainability of the CMake build system.

amstokely committed:
Add `netcdfErrorMessage` Function for Handling NetCDF C++ Exceptions (#33)

* Implemented a function, netcdErrorMessage, to handle NetCDF C++ exceptions.

netcdfErrorMessage prints the netcdf exception's message to the terminal
and returns the corresponding NetCDF error code.

* Reworked netcdfErrorMessage function to provide more info.

* Removed duplicate obs2ioda_cxx_library def

* Removed duplicate definition of obs2ioda_cxx_library.

* Updated formatting so error messages are only two lines.

st-ncar committed:
Make gnssro module private and move derived type definition from subroutine to module.

st-ncar committed:
Move allocation and deallocation of arrays in derived type to separate subroutines

st-ncar committed:
Move netcdf file write to separate subroutine.

st-ncar committed:
Move bufr file read to separate routine and reorganize allocate_gnssro_data_arrays (should have been two separate commits).

st-ncar committed:
Clean up read_write_gnssro subroutine

st-ncar committed:
Improve overall source formatting of gnssro_mod. The hdf5 files generated by the refactored and original gnssro_mod is byte-wise identical () up to here.

amstokely committed:
Add  Fortran Interface for NetCDF File Operations Using C++ API (#34)

* Implemented interface for calling C++ NetCDF's file operations api from fortran.

* Updated createFile to allow the user to set the netcdf filemode.

* Fixed file mode cast error and fileMode 'value' error in the fortran interface.

* Changed add/removeFile methods from int to void.

* Updated add/removeFile docs.

* Updated docs.

st-ncar committed:
Pass correct satellite ID to bending angle observation error calculation.

st-ncar committed:
Add functionality to assign observations to time window.

st-ncar committed:
Add capability to write hourly IODA output files. 6-hourly output files are bit-by-bit identical ($diff) to the ones created with original converter.

amstokely committed:
Add Fortran Support for Creating and Managing NetCDF Groups Using C++ API (#35)

This PR introduces support for creating NetCDF groups and adding them to existing NetCDF files from Fortran using the NetCDF C++ API. The implementation provides a user-friendly Fortran interface that abstracts the complexities of memory management and C/Fortran data type mappings.

Key Features:
NetCDF Group Creation:

Added functionality to create new groups within NetCDF files.
Supports adding groups to specific parent groups or directly to the root group.
User Friendly Fortran Integration:

Memory management is handled entirely within the Fortran interface, reducing the need for user intervention.
Automatic mapping between Fortran strings and null-terminated C strings to ensure compatibility.
Motivation:
This enhancement bridges the gap between Fortran applications and advanced NetCDF capabilities provided by the C++ API. It enables users to work with hierarchical group structures in NetCDF files directly from Fortran without dealing with low-level details.

Next Steps:
Extend the interface to support additional operations:
Adding and retrieving attributes, variables, and data.
Enhanced error reporting and diagnostics.

st-ncar committed:
Add min/max datetime global attributes to h5 output file to better align with other obs.

st-ncar committed:
Merge pull request #38 from st-ncar/bugfix/said

Fix satellite ID bug in gnssro_module

amstokely committed:
Rebased off main. (#36)

st-ncar committed:
Import missing interface in netcdf_cxx_mod required for compilation.

st-ncar committed:
Merge pull request #40 from st-ncar/bugfix/include_netcdf_module

Import missing interface in netcdf_cxx_mod required for compilation.

amstokely committed:
Support for Creating and Setting NetCDF Variables (#39)

* Added support for adding and setting netcdf variables.

* Updated docs.

* Removed template declarations in netcdf_variable.h, updated docs, and removed all getVar related netcdf functions.

* Renamed netcdfPutVar data argument to values.

* Removed @tparam from netcdf variable docstrings and added a default class to setFill/putVar select type statements.

* Removed unused c_f_pointer import

---------

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Support for Creating and Setting NetCDF Attributes in Fortran (#42)

* Added support for writing netcdf attributes using the netcdf-cxx interface.

* Updated PutAtt functions so that they all use the same templated function.

* Removed duplicate defs of setFill interface functions.

* Changed signatures for putAtt functions.

* Generalized template docstrings.

* Updated order of putAtt args in fortran docstring.

st-ncar committed:
Make gnssro_data a local variable.

st-ncar committed:
Encapsulate input arguments related to bufr file into a derived type

amstokely committed:
Ensure `to_c` Returns the C Address of a Trimmed Fortran String   (#45)

* Updated f_c_string_t::to_c function so that it returns the C address of the trimmed input fortran string.

* Replaced len(trim(f_string)) with len_trim(f_string) in f_c_string_t_mod.

st-ncar committed:
Merge pull request #41 from st-ncar/feature/gnssro_refactor

Refactor gnssro observation processing module

amstokely committed:
Update `netcdfAddDim` to Return Dimension ID   (#47)

* Added argument to netcdfAddDim functions that stores the new dimension's dimID.

* Increment dimID value in fortran netcdfAddDim function.

* Removed intermediate status variable in netcdfAddDim.

amstokely committed:
Add Support for Writing NetCDF Character Arrays from Fortran via C++ Wrapper   (#46)

* Added support for writing char array netcdf variables.

* Fixed indentation and removed static cast.

amstokely committed:
Add optional `fillValue` argument to `netcdfAddVar`  (#48)

* Added support for setting fill value when defining a netcdf variable

* updated netcdfAddVar docstring.

amstokely committed:
Declared c_netcdfPutVarChar result variable in c_netcdfPutVarChar interface. (#51)

amstokely committed:
Rebased off main. (#50)

Reverted return variable name fix in c_netcdfPutVarChar.

Fixed error in bind C c_netcdfPutVarChar name.

amstokely committed:
Add support for writing NetCDF integer array attributes (#49)

* Added support for writing int array attributes.

Added support for writing int array attributes.

Cleaned up netcdf attribute code.

* Removed unused f_c_string_attValue in netcdfPutAttArray.

* Fixed formatting.

* Fixed formatting.

st-ncar committed:
Clean up gnssro source file after rebase

st-ncar committed:
Introduce output_info_type to simplify argument list in gnssro subroutines.

st-ncar committed:
Improve comments and variable names

st-ncar committed:
Add support for single-precision floating point array attributes.

st-ncar committed:
Merge pull request #52 from NCAR/feature/gnssro_split

Enable hourly IODA file output for GNSSRO obs

st-ncar committed:
Merge pull request #53 from NCAR/feature/attr_real_array

Add support for single-precision floating point array attributes.

amstokely committed:
Refactor ncio_mod module to use NetCDF C++ wrappers (#43)

Summary
This PR refactors the ncio_mod module by replacing direct NetCDF Fortran calls with a NetCDF C++ wrapper.

Key Changes
NetCDF I/O Update: Replaced all NetCDF Fortran I/O calls with equivalent calls from the NetCDF C++ wrapper (netcdf_cxx).

st-ncar committed:
Use new C++ NetCDF interface in GNSSRO obs converter (#55)

* Replace netcdf calls with new interface.

* Remove commented out calls to old netcdf interface

* Remove legacy group and variable identifiers

* Update call signature of netcdfAddDim to include dimension id.

* Add integer array attributes to IODA file.

* Clean up after rebase off updated main

* Add floating-point array attributes to gnssro ioda file

* Move fill value definition from separate routine to variable definition routine.

amstokely committed:
Enable Cross-Platform Support for FindNetCDF.cmake (#54)

* Rewrote FindNetCDF.cmake so it works on MacOs and Linux.

* Fixed NetCDF_Found logic

amstokely committed:
Define Global Variables for NetCDF Dimensions (#57)

* Created qa branch.

* Create global vars for all dims.

* Create global var for each dim.

* Reverted to original logic for creating nchans dim var.

* Added ws around if statement ( and ).

* Fixed dimension variable name error in write_obs.

* Simplified netcdfPutAtt call in write_obs.

* Updated logic for string array variables that formerly relied on the variable being an nf90_char.

* Fixed coordinate variable dimension name argument bug.

Rebased on `main` to undo changes in `hsd.f90`, `prepbufr_mod.f90`, `radiance_mod.f90`, and `satwnd_mod.f90`, which came from the `feature/char_to_string_netcdf_vars` branch.

amstokely committed:
Change Scan Position Variable Type from Float to Int for IODA v2 Compliance (#58)

amstokely committed:
Refactored Obs2Ioda_CompilerFlags.cmake to only use the mcmedium GNU option when building on Linux, as this flag is not defined by Apple Clang. (#59)

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Convert NetCDF Char Arrays to String Arrays for IODA v2 Compliance   (#56)

* Move scan_position type conversion to ncio_mod.

parent 1a1e9b256b883a05f485451639b3a28b88423b9e
This PR updates NetCDF variable handling by converting character array variables to string array variables, ensuring compliance with the IODA v2 format.

Changed datetime variable type from char to string.

* Fixed allocation error.

* Fixed deallocation bug.

* Changed datetime back to being a nf90_char vs nf90_string.

* Moved all logic related to converting nf90_char to nf90_string variables to ncio_mod. This limits all code changes to within ncio_mod.

amstokely committed:
Fixed c_loc usage bug in netcdf_cxx_mod.f90 (#64)

amstokely committed:
Created a cmake function for registering C++ test executables with CMake. Also configured CMake to build googletest. (#61)

amstokely committed:
Add yaml-cpp as a Dependency via CMake FetchContent (#62)

* Configured CMakeLists.txt in obs2ioda/src/cxx to build yaml-cpp.

* Change yaml-cpp tag from master to 0.8.0 and added include(FetchContent) call to root CMakeLists.

amstokely committed:
Added global variable for location dimension in gnssro_mod. (#66)

amstokely committed:
Add `IodaObsSchema` Class for Validating `obs2ioda` NetCDF Output (#60)

* This PR introduces the `IodaObsSchema` class, which parses the `ObsSpace.yaml` IODA format validation file and uses its contents to validate `obs2ioda` NetCDF output files against the latest IODA format naming and formatting conventions.

Key features of `IodaObsSchema`:
- Ensures `obs2ioda` IO remains compliant with evolving IODA naming conventions.
- Allows `obs2ioda` to continue using its original IO logic while benefiting from schema-based validation.
- Helps maintain consistency with the latest IODA standards without requiring frequent manual updates to `obs2ioda` output logic.

These changes were tested by generating observation files that successfully passed the **3DEnVar** and **HofX** tutorial calculations.

Temp removed tests.

* Added tests for IodaObsSchema class.

* Added test subdirectory to root CMakeLists.txt

* Updated C++ standard. This required for building yaml-cpp on a Mac.

* Removed duplicate yaml-cpp fetch content call.

* Removed duplicate yaml-cpp fetch content call.

* Set C++ standard to 17. This is required for building yaml-cpp on a Mac.

* Moved def of IODA_SCHEMA_YAML to ioda_obs_schema.h.

* Removed duplicate FetchContent(googletest) and include(FetchContent) calls.

* Created macro, set in cmake, that stores the obs2ioda root directory path. This is used to define the ObsSpace.yaml path.

* Moved constructor def for IodsObsVariable out of header.

* Created a config header for setting the ioda schema yaml path at compile time.

This approach avoids the use of macros and will produce a compile vs. runtime error if the ioda schema file does not exist.

---------

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Update obs2ioda to v3 naming and documentation (#71)

* Deleted GMake Makefile.

* Updated obs2ioda executable name to obs2ioda_v3.

* Updated a bunch of names.

* Updated README

* Updated README links.

* Update CMakeLists.txt

* Update CMakeLists.txt

* Update README.md

* Update README.md

* Update README.md

* Update README.md

amstokely committed:
### Add Validation Test Suite for obs2ioda (#73)

* Added support for getting attribute data from netcdf files.

* Created pytest validation test suite.

* Revereted changes in c++ test files and netcdf attributes file.

* Revereted changes in c++ test files and netcdf attributes file.

* Updated README

* Created CMake config files that generate pytest.ini and a python test env file that are used for setting validation test related file paths at compile time.

This allows the validation tests to be run from the build directory and also allow you to define the build directory outside of the obs2ioda source directory.

* Updated README test instructions.

* Updated README and improved test logging.

---------

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Add support for writting IODA V3 format from GOES_ABI converter. (#79)

* GOES-ABI now writes v3!

* Moved all VarMetaData variables to MetaData.

* Moved goes_abi code to obs2ioda-v3 directory.

* Moved all goes_abi code I changed to a different subroutine to ensure I didn't break anything.

* Made building the goes_abi converter optional.

* Moved all obs2ioda code into root src directory. Also added tests for all goes_abi changes.

* Updated test structure and changed datetime type from string long int in goes_abi converter.

* Updated gitignore.

* Updated marker descriptions.

* Added docs to all new test code.

---------

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Changed the output_dir length to 512 in the write_goes_abi_ioda_v3 test. (#80)

This is required for directory paths that exceed 100 (previous size) characters.

amstokely committed:
Added global variables for each dimension. (#82)

This PR updates the write_iodav3_netcdf subroutine to write explicit global variables for each dimension in the output NetCDF file. This change ensures compatibility with the IODA v3 format, which requires dimension variables to be present at the global level.

amstokely committed:
Update README with GOES-ABI Converter Installation and Usage Instructions (#83)

Summary:
This PR updates the README.md to document how to install and use the GOES-ABI converter included in obs2ioda. It also improves the instructions for running the validation test suite.

Changes:

Added instructions for enabling the GOES-ABI converter during the CMake configuration step (-DBUILD_GOES_ABI_CONVERTER=ON).
Added a section explaining how to run the GOES-ABI converter from the command line.
Expanded the test instructions to show how to run specific test suites using pytest -m, including the goes_abi marker.

amstokely committed:
Deleted obsolete comment in README (#85)

amstokely committed:
Update `datetime` handling to use `dateTime` with correct type and fill behavior and simplify environment setup (#86)

* Added global variables for each dimension. (#82)

This PR updates the write_iodav3_netcdf subroutine to write explicit global variables for each dimension in the output NetCDF file. This change ensures compatibility with the IODA v3 format, which requires dimension variables to be present at the global level.

Update README with GOES-ABI Converter Installation and Usage Instructions (#83)

Summary:
This PR updates the README.md to document how to install and use the GOES-ABI converter included in obs2ioda. It also improves the instructions for running the validation test suite.

Changes:

Added instructions for enabling the GOES-ABI converter during the CMake configuration step (-DBUILD_GOES_ABI_CONVERTER=ON).
Added a section explaining how to run the GOES-ABI converter from the command line.
Expanded the test instructions to show how to run specific test suites using pytest -m, including the goes_abi marker.

Changed name of datetime variable to dateTime and fixed bug that occurred when writing dateTime with superobs enabled in goes_abi.

Defined linker language for fortran test targets.

Changed netcdf4 to netCDF4 in README instructions.

Fixed how Channel variable is set in goes_abi converter. Also added a function for creating fortran ctest tests.

Added units attribute for dateTime variable.

Updated test reference files.

Updated goes_abi ouput file naming.

Added docstring for  set_goes_abi_out_fname subroutine.

Added a conda env recipe yaml for creating the obs2ioda python testing env.

Also updated the README.

Created csh derecho environment scripts along with a bash intel derecho env script.

* Deleted leftover commit tag from rebase in README.

amstokely committed:
Fix integer truncation bug in superob BT calculation by refactoring to pure function (#90)

* parent dacce1e0ae5b6330187c522b8fd4807930d1ab0c
author Andy Stokely <amstokely> 1749142514 -0700
committer amstokely <andystokely@gmail.com> 1749842995 -0600

Changed name of datetime variable to dateTime and fixed bug that occurred when writing dateTime with superobs enabled in goes_abi.

Changed netcdf4 to netCDF4 in README instructions.

Fixed how Channel variable is set in goes_abi converter. Also added a function for creating fortran ctest tests.

Added units attribute for dateTime variable.

Updated test reference files.

Added docstring for  set_goes_abi_out_fname subroutine.

Fixed bug in read_HSD releated to the intermediate variable tb being an integer instead of a real.

Fixed bug in read_HSD releated to the intermediate variable tb being an integer instead of a real.

* Removed unused variable in hsd.f90

* Updated README

* Removed outdated build instructions.

* Updated HSD FLDK file conversion instructions.

* Updated HSD FLDK conversion instructions.

* Cleaned up C++ test names.

---------

Co-authored-by: Andy Stokely <amstokely>

amstokely committed:
Fix Incorrect Solar Zenith Angle Calculation for AHI Conversions (#96)

* Fixed solar zenith angle bug and added unit tests for calc_solar_zenith_angle
* Initialized ihh.

zhumingying committed:
update obs longitude from 0-360 to -180 to 180, and adding reading cris_n21 with obserror (#97)

(1) Update observation longitudes from the (0, 360) range to (-180, 180)
(2) Add support for reading CrIS_N21 data with observation error

BenjaminRuston committed:
integrate updated obs2ioda from NCAR into JCSDA internal for unified repository

BenjaminRuston committed:
disable LFS tracking for 'write_goes_abi_ioda_v3.nc'

BenjaminRuston commented:
@amstokely there was a file it complained about as the git-lfs tracking wasn't correct:
```
Encountered 1 file that should have been a pointer, but wasn't:
	src/ncar/obs2ioda/test/validation/data/goes_abi/write_ioda_v3/reference/write_goes_abi_ioda_v3.nc
```

for now I set it to ignore git-lfs by altering the `.gitattributes` but think we can fix this with cleanup of the branch now that it's been moved to internal

amstokely committed:
Deleted validation tests for ncar obs2ioda.

These are no longer needed as ioda_converters has a robust CI/CD pipeline in place.

amstokely committed:
Moved netcdf C++ unit tests into test suite.

amstokely committed:
Deleted pytest cmake config files.

ibanos90 commented:
Hi @BenjaminRuston, thank you for your efforts on this! 

As we work toward the unification of the code under ioda-converters, I wanted to highlight that, historically, obs2ioda has been very easy to compile with minimal dependencies. It would be ideal to preserve this simplicity. 

My question is, would it be possible to add an option to compile obs2ioda as a standalone component? Given that obs2ioda do not depend on any other ioda/oops code, could we add a CMake argument to enable standalone compilation?

BenjaminRuston commented:
@ibanos90 thank you and I am not opposed, there is another issue open where we're trying to help someone build the ioda-converters from the public jedi-bundle.  Here we recommended to comment out a block of lines specifically these lines 60-73:
https://github.com/JCSDA/jedi-bundle/blob/47c9177d337cd2a8426eb09e1158679b7379cead/CMakeLists.txt#L60-L73

would that be acceptable to put into a conditional driven by a CMake argument

ibanos90 commented:
> @ibanos90 thank you and I am not opposed, there is another issue open where we're trying to help someone build the ioda-converters from the public jedi-bundle. Here we recommended to comment out a block of lines specifically these lines 60-73: https://github.com/JCSDA/jedi-bundle/blob/47c9177d337cd2a8426eb09e1158679b7379cead/CMakeLists.txt#L60-L73
> 
> would that be acceptable to put into a conditional driven by a CMake argument

@BenjaminRuston I believe this will still build the whole ioda-converters, correct?
Would it be possible to build just obs2ioda, without having to compile the rest of the ioda-converters code? I guess with current capabilities it may not be possible, right?

amstokely committed:
Updated unit tests and fixed memory leak in f_c_string code.

BenjaminRuston commented:
> Would it be possible to build just obs2ioda, without having to compile the rest of the ioda-converters code?

hmmm, I'm not sure about that one, and yes @ibanos90, you're correct something that would skip L60-L73 in the CMakeLists.txt example above would still make all of ioda-converters

when you make your bundle sounds like you typically don't make ioda-converters then... let me have a think on this and see if have any other insights

amstokely committed:
ncar obs2ioda now compiles, but the cmake rpath is incorrectly set :(

amstokely committed:
Deleted f_c_)string_1D_t_mod.f90 as it has been replaced by the memory safe f_c_string_array_t module.

amstokely committed:
Fixed ncar obs2ioda linking problems.

Still have to add <build dir>/lib to ld library path to run ncar obs2ioda unit tests.

amstokely commented:
> > Would it be possible to build just obs2ioda, without having to compile the rest of the ioda-converters code?
> 
> hmmm, I'm not sure about that one, and yes @ibanos90, you're correct something that would skip L60-L73 in the CMakeLists.txt example above would still make all of ioda-converters
> 
> when you make your bundle sounds like you typically don't make ioda-converters then... let me have a think on this and see if have any other insights

@ibanos90 Ideally, it would be nice to retain the ability to build only obs2ioda. However, I don’t think that’s feasible anymore since obs2ioda now relies on ioda, particularly for access to the ObsSpace.yaml. I previously kept a copy of that file within obs2ioda, but now that it's part of a library with a direct dependency on ioda, maintaining a separate copy doesn't make much sense.

ibanos90 commented:
> > > Would it be possible to build just obs2ioda, without having to compile the rest of the ioda-converters code?
> > 
> > 
> > hmmm, I'm not sure about that one, and yes @ibanos90, you're correct something that would skip L60-L73 in the CMakeLists.txt example above would still make all of ioda-converters
> > when you make your bundle sounds like you typically don't make ioda-converters then... let me have a think on this and see if have any other insights
> 
> @ibanos90 Ideally, it would be nice to retain the ability to build only obs2ioda. However, I don’t think that’s feasible anymore since obs2ioda now relies on ioda, particularly for access to the ObsSpace.yaml. I previously kept a copy of that file within obs2ioda, but now that it's part of a library with a direct dependency on ioda, maintaining a separate copy doesn't make much sense.

I see, thanks @amstokely

amstokely committed:
Added integration tests for ncar obs2ioda.

The tests are the same ones used for bufr2nc_fortran. Since obs2ioda writes v3 vs v2, the tests currently fail.

amstokely committed:
Updated reference value in set_goes_abi_out_fname test.

amstokely committed:
Updated CI/CD

fabiolrdiniz commented:
Related to @ibanos90's point, would it be a problem to have a makefile inside the folder? And/or the instructions inside the README file on how to build a standalone version of the converter? I'm more inclined towards the latter. Regardless of the chosen option, it would be helpful to have some information about these details included in the README file.

ibanos90 commented:
> Related to @ibanos90's point, would it be a problem to have a makefile inside the folder? And/or the instructions inside the README file on how to build a standalone version of the converter? I'm more inclined towards the latter. Regardless of the chosen option, it would be helpful to have some information about these details included in the README file.

That would be great! Thanks for looking into this, @fabiolrdiniz!

amstokely committed:
Removed old test directory for ncar obs2ioda.

amstokely committed:
Deleted unused cmake files in ncar/obs2ioda repo.

amstokely committed:
Fixed bug that appears in apple clang 15 when writing global netcdf components in ncar obs2ioda.

amstokely committed:
Re-added ncar obs2ioda cmake lists.

amstokely committed:
Removed new obs2ioda tests temp.

amstokely committed:
Fixed yaml cpp linking problem.

amstokely committed:
Fixed yaml-cpp linking issue.

amstokely commented:
@eap This PR introduces  `yaml-cpp` as a dependency, which is built via `FetchContent`. Since this does not require users to install an additional dependency, do you see this being a problem?

amstokely committed:
Refactored IodaObsSchema to make it easier to add a eckit yaml interface in the future.

amstokely committed:
Fixed compilation bug due to netcdf_utils.h not including the <memory> header.

BenjaminRuston committed:
Merge branch 'develop' into feature/obs2ioda_alignment

amstokely committed:
Added f_c_string memory allocation check.

BenjaminRuston commented:
@amstokely needs reviews to merge!  but this appears ready to go (for this initial version)

BenjaminRuston commented on src/ncar/obs2ioda/src/cxx/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+FetchContent_Declare(
+        yaml-cpp
+        GIT_REPOSITORY https://github.com/jbeder/yaml-cpp.git
+        GIT_TAG 0.8.0
```
@eap this is  a new dependency, as believe the eckit handle the yaml parsing

not this is a first implementation of this new `obs2ioda` and would be nice if can make this common.

amstokely replied:
While I'm not familiar with eckit, it should be straightforward for someone who is to replace the YAML parsing code with an eckit-based interface equivalent to the current [yaml-cpp interface](https://github.com/JCSDA-internal/ioda-converters/blob/de6cb65060b755afb6646d9a61cece0cc183f1f1/src/ncar/obs2ioda/src/cxx/ioda_obs_schema.h#L20).

climbfuji replied:
I checked the license, yaml-cpp has an MIT license and it can be used in an Apache 2 licensed product like JEDI

BenjaminRuston commented:
@eap added you mostly if you could scan the CMake files, and make sure nothing may cause issues

this brings this converter inline with IODAv3 which is great, and we expect there will likely be some additional changes needed on top of this but any major roadblocks can be addressed now and other issues let's tackle in smaller chunks if you agree

amstokely commented:
The new [C++](https://github.com/NCAR/obs2ioda/tree/feature/obs2ioda_alignment/test/cxx) and[ Fortran ](https://github.com/NCAR/obs2ioda/tree/feature/obs2ioda_alignment/test/fortran)wrapper I/O is very well tested. The tests are not included in this PR as they introduce more dependencies (`googletest` and `pFUnit`) that are outside the JEDI ecosystem.

climbfuji commented on src/ncar/obs2ioda/.gitignore:
```diff
@@ -0,0 +1,7 @@
+**/__pycache__
+test/validation/test_utils/FilePathConfig.py
+.obs2ioda
+build
```
```suggestion
build

```

climbfuji commented on src/ncar/obs2ioda/cmake/Obs2Ioda_CompilerFlags.cmake:
```diff
@@ -0,0 +1,36 @@
+# -fpe0: Stop execution when a floating-point exception occurs
+set(FORTRAN_COMPILER_INTEL_DEBUG_FLAGS
+    $<$<COMPILE_LANGUAGE:Fortran>:-check uninit -ftrapuv -g -traceback -fpe0>
+)
```
```suggestion
)

```

climbfuji reviewed (commented):
What was the reasoning for adding this de-facto separate piece of software to ioda-converters with all the complications for the build system etc?

Why not make ncar obs2ioda its own package/repository?

snyderdad commented:
@climbfuji : That approach (own repo for obs2ioda) would be OK w/ NCAR.  Perhaps preferable if it eases or enables a stand-alone build.

climbfuji commented:
> @climbfuji : That approach (own repo for obs2ioda) would be OK w/ NCAR. Perhaps preferable if it eases or enables a stand-alone build.

Modularity certainly helps with the spack-ification of the JEDI packages, and looking at how complicated the build integration between ioda-converters (ecbuild) and ncar obs2ioda (standard cmake) is, I think it would be better altogether.

amstokely commented:
> > @climbfuji : That approach (own repo for obs2ioda) would be OK w/ NCAR. Perhaps preferable if it eases or enables a stand-alone build.
> 
> Modularity certainly helps with the spack-ification of the JEDI packages, and looking at how complicated the build integration between ioda-converters (ecbuild) and ncar obs2ioda (standard cmake) is, I think it would be better altogether.


@BenjaminRuston @climbfuji @snyderdad  If it’s important for JCSDA to maintain `obs2ioda` within `ioda-converters`, it should be relatively straightforward to convert its CMake build system to use `ecbuild`. While I don’t have time to do this myself, it would likely take only a few days of work for someone on the `ioda` team to make the necessary changes.

BenjaminRuston commented:
> @BenjaminRuston @climbfuji @snyderdad If it’s important for JCSDA to maintain obs2ioda within ioda-converters, it should be relatively straightforward to convert its CMake build system to use ecbuild. While I don’t have time to do this myself, it would likely take only a few days of work for someone on the ioda team to make the necessary changes.

My strategy will be to open an issue to do the conversion and since this would be a different person would propose that work be done in a separate issue and PR

@climbfuji or @eap is there any unforeseen build issues this would cause for other broader users if not this can continue from this initial contribution

climbfuji commented:
> > @BenjaminRuston @climbfuji @snyderdad If it’s important for JCSDA to maintain obs2ioda within ioda-converters, it should be relatively straightforward to convert its CMake build system to use ecbuild. While I don’t have time to do this myself, it would likely take only a few days of work for someone on the ioda team to make the necessary changes.
> 
> My strategy will be to open an issue to do the conversion and since this would be a different person would propose that work be done in a separate issue and PR
> 
> @climbfuji or @eap is there any unforeseen build issues this would cause for other broader users if not this can continue from this initial contribution

I think this is going into the wrong direction. Instead of converting standard cmake to ecbuild, we should be doing the opposite ... but that's just me. That being said, both ioda-converters and obs2ioda being ecbuild is still cleaner than a mix. Most important is that all source code for obs2ioda is included in the ioda-converters repository and there are no "hidden" downloads of code during a `cmake`, `ecbuild`, or `make` step.

amstokely commented:
> > > @BenjaminRuston @climbfuji @snyderdad If it’s important for JCSDA to maintain obs2ioda within ioda-converters, it should be relatively straightforward to convert its CMake build system to use ecbuild. While I don’t have time to do this myself, it would likely take only a few days of work for someone on the ioda team to make the necessary changes.
> > 
> > 
> > My strategy will be to open an issue to do the conversion and since this would be a different person would propose that work be done in a separate issue and PR
> > @climbfuji or @eap is there any unforeseen build issues this would cause for other broader users if not this can continue from this initial contribution
> 
> I think this is going into the wrong direction. Instead of converting standard cmake to ecbuild, we should be doing the opposite ... but that's just me. That being said, both ioda-converters and obs2ioda being ecbuild is still cleaner than a mix. Most important is that all source code for obs2ioda is included in the ioda-converters repository and there are no "hidden" downloads of code during a `cmake`, `ecbuild`, or `make` step.

@climbfuji I was suggesting converting the cmake build to ecbuild solely for consistency reasons, not because I like ecbuild :). Without using `FetchContent`, wouldn't that mean that all `ioda-converter` users would have to install `yaml-cpp` in their JEDI environment? `yaml-cpp` is very easy to install, but I'm always hesitant to require people to manually install a new library that is outside the JEDI spack-stack environment.

climbfuji commented:
> > > > @BenjaminRuston @climbfuji @snyderdad If it’s important for JCSDA to maintain obs2ioda within ioda-converters, it should be relatively straightforward to convert its CMake build system to use ecbuild. While I don’t have time to do this myself, it would likely take only a few days of work for someone on the ioda team to make the necessary changes.
> > > 
> > > 
> > > My strategy will be to open an issue to do the conversion and since this would be a different person would propose that work be done in a separate issue and PR
> > > @climbfuji or @eap is there any unforeseen build issues this would cause for other broader users if not this can continue from this initial contribution
> > 
> > 
> > I think this is going into the wrong direction. Instead of converting standard cmake to ecbuild, we should be doing the opposite ... but that's just me. That being said, both ioda-converters and obs2ioda being ecbuild is still cleaner than a mix. Most important is that all source code for obs2ioda is included in the ioda-converters repository and there are no "hidden" downloads of code during a `cmake`, `ecbuild`, or `make` step.
> 
> @climbfuji I was suggesting converting the cmake build to ecbuild solely for consistency reasons, not because I like ecbuild :). Without using `FetchContent`, wouldn't that mean that all `ioda-converter` users would have to install `yaml-cpp` in their JEDI environment? `yaml-cpp` is very easy to install, but I'm always hesitant to require people to manually install a new library that is outside the JEDI spack-stack environment.

That raises another question. Since JEDI uses `eckit` for parsing yaml files, does it make sense to introduce another dependency for doing the same task? I understand that while obs2ioda was developed in isolation outside of the JEDI system, using `yaml-cpp` made a lot of sense. But now that the `obs2ioda` code is integrated in the JEDI codebase, wouldn't it make more sense to use `eckit`?

BenjaminRuston commented:
believe we are largely in agreement here.  for me still happy for suggestions on order of things.

there is a bit of a handoff as it goes to the next stage, which is the `converting standard cmake to ecbuild`.  This then also brings the ability to remove `yaml-cpp`.

I am panhandling for 2cents @eap and @climbfuji if it is too messy to keep that dependency now and to execute the next stage subsequently? 

With the `FetchContent` to be deprecated once verified can use other pre-existing method (withholding extra discussion)

amstokely commented:
> believe we are largely in agreement here. for me still happy for suggestions on order of things.
> 
> there is a bit of a handoff as it goes to the next stage, which is the `converting standard cmake to ecbuild`. This then also brings the ability to remove `yaml-cpp`.
> 
> I am panhandling for 2cents @eap and @climbfuji if it is too messy to keep that dependency now and to execute the next stage subsequently?
> 
> With the `FetchContent` to be deprecated once verified can use other pre-existing method (withholding extra discussion)

@climbfuji @BenjaminRuston Updating the YAML parsing code to use eckit instead of yaml-cpp should be fairly straightforward for someone familiar with eckit (I’m not). The key step would be implementing an interface equivalent to the current yaml-cpp one used [here](https://github.com/JCSDA-internal/ioda-converters/blob/de6cb65060b755afb6646d9a61cece0cc183f1f1/src/ncar/obs2ioda/src/cxx/ioda_obs_schema.h#L20).

The NCAR repo includes a [Googletest-based test suite](https://github.com/NCAR/obs2ioda/blob/feature/obs2ioda_alignment/test/cxx/ioda_obs_schema.test.cc) for the YAML parsing code that could help validate the new implementation. If the test suite is to be used in production under eckit, it would need to be adapted accordingly.

srherbener commented:
Here's how I see this situation. The main issue that needs to be fixed is that we currently have (at least) 4 evolutionary copies of the "obs2ioda" converter, starting with the original Jamie Bresch converter along with various snapshots and ending with Andy's obs2ioda converter. The goal is to obsolete the older snapshots and move forward with only Andy's obs2ioda converter. The existence of 4 or more snapshots of the same converter is not supportable and maintainable.

I see two options:
1. Keep the converter in NCAR which means that NCAR is responsible for its maintenance and development
2. Move the converter to JEDI (ioda-converters repo) which means that JCSDA and partners are responsible for its maintenance and development

I think the deciding factor for picking one of the options, is how important it is to keep this converter stand alone. If that is the highest priority then the converter should stay in NCAR (option 1). You would be free to continue with cmake, and use whatever library dependencies you see fit.

We started this PR because we were assuming that the highest priority was to consolidate the effort to support and develop this converter by moving it to JEDI, with the hope of spreading out the responsibilities (among the several people that are supporting the various snapshots) because there is too much demand on Andy's time. If this is the correct priority, then we should continue with option 2. However, with option 2, we need to pull obs2ioda into the JEDI system meaning that we should not introduce new library dependencies for functionality we already have in spack-stack. This means that we should convert obs2ioda to a oops::Application, which will facilitate using eckit::Configuration, and we should use the eckit/fckit test framework, etc.

This also means that we shouldn't force it to be a standalone converter. You would be able to build it by cloning jedi-bundle, and commenting out the ecbuild_bundle commands for every package except oops, ioda, ioda-data and ioda-converters. This will still build quickly.

Also, ecbuild is simply a set of cmake macros, and it runs cmake under the hood. It may not be all that much work (if any) to build with ecbuild instead of cmake. This is primarily due to the fact that the top-level CMakeLists.txt file is already in ioda-converters and that is where the bulk of "ecbuidl" flavor configuration resides. 

If we agree that option 2 is the right choice, then I would advocate for biting the bullet and doing the conversion of obs2ioda into a JEDI (OOPS) style application, etc from the start. I think that might actually turn out to be less effort than trying to add new dependencies (and manage those dependencies) to spack-stack.

What do others think about this?

BenjaminRuston commented:
@srherbener we are going with option2,

think we've been using option1 for the last few years and have decided it's not in everyone's best interest.  This has been discussed with the MMM group and we have agreement.  

The decision now is move towards ecbuild from standard CMake, and this will also allow removal of yaml-cpp dependency

ability to allow for a stand-alone build could be possible most simply with the addition of a Makefile which would not be executed by majority of partners but could allow a build of the single component.  This is something that is not encouraged broadly only for special circumstances, we are deeming so in this case

BenjaminRuston commented:
thanks @rajichidamb , and yes let's proceed with current track which is to have this in parallel with the other version

for now the `FetchContent` is done so that all ioda-converter users do not have to install yaml-cpp independently in their JEDI environment

this work will be picked up by JCSDA to do the `ecbuild` conversion of the CMake and then we can proceed with a few leftover bits: 
- verifying this against the to be deprecated `bufr2nc_fortran.x`
- making sure one more time bug-fixes from MMM and JCSDA are incorporated
- method for stand-alone build,, currently plan is a largely to be ignored Makefile in the directory with appropriate modifications to documentation about why it's there and how to use it -- default ignore it

BenjaminRuston committed:
Merge branch 'develop' into feature/obs2ioda_alignment

fcvdb commented:
I built the code on Hercules (GNU spack-stack 1.9) and ran the converters (`bufr2nc_fortran.x prepbufr.gdas.2021021600.nr`) on this prepbufr file (from NCAR RDA derecho: `/glade/campaign/collections/rda/data/d337000/prepnr/2021/prepbufr.gdas.2021021600.nr`). The converter built and completed. 

Using `ncdiff`, I compared the generated IODA files (`obs.aircraft_obs_2021021600.nc4, obs.ascat_obs_2021021600.nc4, obs.radiosonde_obs_2021021600.nc4, obs.satwind_obs_2021021600.nc4`) withe same files generated from the develop branch. The files are identical.

fcvdb reviewed (approved):
Thanks Ben!

fabiolrdiniz commented on src/ncar/obs2ioda/cmake/Obs2Ioda_CompilerFlags.cmake:
```diff
@@ -0,0 +1,36 @@
+    $<$<COMPILE_LANGUAGE:Fortran>:-ffree-line-length-none>
+)
+
+# Set Fortran compiler flags specific to the GNU Compiler and Linux OS.
+# -mcmodel=medium: Allow for larger datasets in memory
+set(FORTRAN_COMPILER_GNU_LINUX_FLAGS
+        $<$<COMPILE_LANGUAGE:Fortran>:-mcmodel=medium>
+)
```
I'm seeing an error message related to this change: https://github.com/JCSDA-internal/ioda-converters/issues/1682
